backend/monitor/supervisor.go

```diff
@@ -146,38 +178,37 @@ func (m *EtcdSupervisor) getMonitor(ctx context.Context, key string) (*monitor,
 // is witnessed, it calls the provided HandleFailure func. If a PUT event is
 // witnessed, the watcher is stopped.
 func watchMon(ctx context.Context, cli *clientv3.Client, mon *monitor, failureHandler func(), shutdownHandler func()) {
+	if mon.ttl == 0 {
+		panic("zero-duration ttl")
+	}
+	ctx, cancel := context.WithCancel(ctx)
 	responseChan := cli.Watch(ctx, mon.key)
+	timer := time.NewTimer(time.Duration(mon.ttl) * time.Second)
 	go func() {
-		for wresp := range responseChan {
-			for _, ev := range wresp.Events {
-				if ev.Type == mvccpb.DELETE {
-					failureHandler()
-					return
-				}
-				// if there is a PUT on the key, the lease has been extended,
-				// and we want to kill this watcher to avoid duplicate watchers.
-				if ev.Type == mvccpb.PUT {
-					shutdownHandler()
-					return
+		defer timer.Stop()
+		defer cancel()
+		for {
+			select {
+			case wresp := <-responseChan:
+				for _, ev := range wresp.Events {
+					if ev.Type == mvccpb.DELETE {
+						failureHandler()
+						return
+					}
+					// if there is a PUT on the key, the lease has been extended,
+					// and we want to kill this watcher to avoid duplicate watchers.
+					if ev.Type == mvccpb.PUT {
+						shutdownHandler()
+						return
+					}
 				}
+			case <-timer.C:
+				failureHandler()
+				return
 			}
 		}
 	}()
-
-	sigs := make(chan os.Signal, 1)
-	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
-	go func() {
-		sig := <-sigs
-
-		logger.Debugf("signal %s received, revoking lease for key %s",
-			sig.String(),
-			mon.key,
-		)
-
-		if _, err := cli.Lease.Revoke(ctx, mon.leaseID); err != nil {
-			logger.WithError(err).Warningf("could not revoke the lease %v for the key %s",
-				mon.leaseID, mon.key,
-			)
-		}
-	}()
+	leasesMu.Lock()
+	defer leasesMu.Unlock()
```

Comment:
My 2 cents on style related to critical sections: I prefer to explicitly call `Unlock()` as soon as I'm done with a resource, rather than say the equivalent of "from this point on, hold the lock until this function returns".

You don't want someone else potentially unknowingly adding code in a critical region!

Reply:
In this case I agree with you. However in other cases, it is a very good practice to defer the Unlock: if a panic occurs in the CS, then the mutex will still be unlocked. This can make a big difference in code that calls recover().